Improve attr mismatch msg

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149576

Differential Revision: [D71513041](https://our.internmc.facebook.com/intern/diff/D71513041)

tugsbayasgalan committed:
Improve attr mismatch msg

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71513041).

avikchaudhuri commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1659,8 +1659,27 @@ def _collect_assigned_tensor_attributes(kp, v, _v):
+        if len(new_attrs) != len(snapshot):
+            newly_created_attrs = []
+            deleted_attrs = []
+            for k in new_attrs:
```
isn't there a better way to do set diff lol

avikchaudhuri commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1659,8 +1659,27 @@ def _collect_assigned_tensor_attributes(kp, v, _v):
+                    deleted_attrs.append(k)
+
+            raise ValueError(
+                f"During export following attrs are deleted: {deleted_attrs} "
```
Eh, we should tailor the error message when one of these is empty. Otherwise the message for the test is pretty weird (nothing was deleted). Frankly deleting would be pretty uncommon...

Also, say "the following attributes were" instead of "following attrs are"

tugsbayasgalan committed:
Update on "Improve attr mismatch msg"


Differential Revision: [D71513041](https://our.internmc.facebook.com/intern/diff/D71513041)

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1659,8 +1659,27 @@ def _collect_assigned_tensor_attributes(kp, v, _v):
+
+            if len(added_attrs) > 0:
+                raise ValueError(
+                    f"During export following attrs were created: {added_attrs} "
```
I think there are two things here that seem ambiguous:

(1) whether or not the user puts an attribute on their `nn.Module` at init time, or inside of their model forward

(2) whether or not the attribute that the user sets is properly a buffer

I think this PR is mainly addressing (1). But is the buffer part necessary? We should probably make both of these explicit in the message (they should set the attribute at model init time and not in the forward, **and** they should set it through the `register_buffer` API)

tugsbayasgalan replied:
Sounds good!

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71513041).

tugsbayasgalan committed:
Update on "Improve attr mismatch msg"


Differential Revision: [D71513041](https://our.internmc.facebook.com/intern/diff/D71513041)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71513041).

tugsbayasgalan committed:
Update on "Improve attr mismatch msg"


Differential Revision: [D71513041](https://our.internmc.facebook.com/intern/diff/D71513041)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71513041).

avikchaudhuri commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1659,8 +1659,29 @@ def _collect_assigned_tensor_attributes(kp, v, _v):
+
+            if len(added_attrs) > 0:
+                raise ValueError(
+                    f"During torch.export, following attrs were created in the model.forward: {added_attrs} "
```
`following attrs were created in the model.forward` -> `the following attributes were created`?
